Add bidirectional isolation

eemeli opened the pull request:
This fulfills the spec requirements for [Handling Bidirectional Text](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#handling-bidirectional-text).

The ApplyBidiIsolation AO will in the spec text need more formal descriptions of how the `"compatibility"` and `"none"` strategies work, but I thought that would be a bit heavy for the README.

We could also define something like a `"best fit"` option to match what's done with locale negotiation, but I left that out as I wasn't sure exactly what it would look like.

CC @aphillips

eemeli commented:
@aphillips Do I understand right that you think we should include `dir` as an explicit option for the formatter in addition to `locale`, and to include it for each of the formatted parts?

My presumption was that from the locale code we could get the script, from which we could determine the direction. But I presume that's not always valid then?

aphillips commented:
@eemeli More like: every formatted part should have a `dir` property. This can be generated by a formatter (and might, indeed, be inferred from the locale if that is how the formatter figures out the base paragraph direction of e.g. a date string). However, not every placeholder is generated by a formatter. Having a `dir` property also allows assembly of messages whose character sequences would fool `auto`. Consider:

Input variables:
```json
{  "item": {
    "id": "978-111887164-5",
    "title": {
        "value": "HTML و CSS: تصميم و إنشاء مواقع الويب!",
        "lang": "ar",
        "dir": "rtl"
    }
}
```

And message:

```
{Your purchased the item {$title.value @lang=$title.lang @dir=$title.dir}.}
```

In this case, you want the title bidi isolated (it's in Arabic) and you don't want a direction to be LTR (implied by `HTML` on the front). Only the metadata associated with the original data value conveys this.

We have a long document discussing this: [Strings on the Web: Language and Direction Metadata](https://www.w3.org/TR/string-meta/). It includes a discussion of why language tags (and their extant or implied script subtag) is only a fallback.

eemeli commented:
After discussion with @aphillips elsewhere, I've rebased and updated this PR to include `dir` as a supported option, and to use that rather than the locale to determine the isolation to use.

This also adds `dir: 'ltr' | 'rtl' | 'auto'` as a property of the formatted parts that also include `locale`.

aphillips commented on README.md:
```diff
@@ -200,11 +239,21 @@ To determine the value `res` returned by the `formatToParts()` method,
-   1. Set `parts` to be `[{ type: 'fallback', locale: 'und', source: mv.source }]`.
+   1. Set `parts` to be `[{ type: "fallback", source: mv.source }]`.
+   1. Set `dir` to be `"auto"`.
+1. Let `bidi` be the `{ start: string, end: string }` result of calling
+   `ApplyBidiIsolation(bidiIsolation, msgDir, dir)`.
+1. If `bidi.start` is not an empty string:
+   1. Append `{ type: 'bidiIsolation', value: bidi.start }` to `res`.
 1. For each `part` or `parts`:
    1. Append `part` to `res`.
+1. If `bidi.end` is not an empty string:
+   1. Append `{ type: 'bidiIsolation', value: bidi.end }` to `res`.
```
This is the discussion we were having on the parts spec in MF2. This design makes it hard for developers who are using the parts to build HTML using HTML's `dir` attribute, because instead of each part just having a direction, the developer getting a `bidiIsolation` part has to peek ahead to figure out what the text is that it applies to. They then need to decide whether to generate *separate* bidi markup (such as a `span` element) or try to apply `dir` to a known inline element (which is why they got parts in the first place--to be able to decorate the string with `em`, `strong`, `bdi` 😉, or some CSS class. 

This algorithm works for `compatibility` mode, since it just has the bidi control in it. Is your intention that `none` conveys what I just described above? If so, then the name might be wrong?

Perhaps make the names: 

`compatibility` => `controls`
`none` => `inline` or maybe `attributes`

eemeli replied:
The `bidiIsolation` parts are _in addition_ to the `dir` that's on the formatted parts. So a developer building HTML could just ignore the `bidiIsolation` parts and build e.g. `<span dir>` elements as necessary from the `dir` attributes.

On the other hand, in this particular case the `bidiIsolation` parts also act as start/end markers in case the value formats to more than one top-level part (Note: this is different from e.g. number parts _containing_ sub-parts), so that a single multi-part value could be completely encompassed in an appropriate `<span dir>` if necessary.

Regarding the "compatibility" name, see my comment [here](https://github.com/tc39/proposal-intl-messageformat/pull/30/files#r1313199545). Regarding "none", I still think it's right to say that using it no bidi isolation is provided. Given that the `dir` values are still included in the parts output, it is of course possible for a consumer to construct its own bidi isolation from the provided information, but Intl.MessageFormat isn't doing it in that case.

aphillips commented on README.md:
```diff
@@ -525,6 +581,7 @@ In such a case, a fallback representation is used instead for the value:
 interface MessageFallback {
   type: 'fallback';
   locale: 'und';
+  dir: 'auto'
```
Both locale and dir can be known from the call to MF. Any reason to suppress that information?

eemeli replied:
For fallback, we can't know without introspection what the `source` direction is, and we should not presume that the contents match the locale. In other words, we do not know the locale or direction of a fallback part.

aphillips replied:
Again, if `source` is externally provided and isn't the pattern string, I can agree about the direction. The locale, though, really is the locale of the formatter, I think. Otherwise the caller won't know if some locale defaulting happened. If the `source` is just an identifier, its actual language is probably "programmer-ese" and not any particular natural language. But the language of the words in the source don't have to be in the locale's language for that to be the correct locale for the message.

aphillips commented on README.md:
```diff
@@ -179,10 +197,31 @@ To determine the value `res` returned by the `format()` method,
+1. Let `dir` be `mv.dir`.
 1. Let `strval` be the result of calling `mv.toString()`.
 1. If the call fails or `strval` is not a string:
    1. Set `strval` to be the concatenation of `{`, `mv.source`, and `}`.
```
Weird nit: the fallback uses enclosing punctuation `{` and `}`, which get special treatment in the bidi algorithm. Since `mv.source` can be a bidirectional string, you probably want to isolate the source string from the brackets. You could do that by putting `\u2068` after `{` and `\u2069` before `}` or by applying the isolation strategy.

Also, the bidi algorithm will pair the enclosing brackets in most cases, so maybe ignore this.

eemeli replied:
The FSI+PDI wrapping is handled by the next line setting `dir=auto` and then `ApplyBidiIsolation()` adding the isolation.

aphillips reviewed (changes requested):
Thanks for these changes. Some comments.

eemeli commented on README.md:
````diff
@@ -145,18 +145,36 @@ for the creation of `MessageFormat` instances.
 
 ```ts
 interface MessageFormatOptions {
+  bidiIsolation?: 'compatibility' | 'none';
````
The "compatibility" name is explicitly mentioned in the MF2 spec: https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#handling-bidirectional-text

If it should be called something else, it would be best to re-name it there so that the same name will end up in other implementations as well.

The bidi isolation strategy is set in the constructor, so each needs to work for all formatting targets.

aphillips replied:
I agree that we should figure out the naming there.

eemeli commented on README.md:
```diff
@@ -179,10 +197,31 @@ To determine the value `res` returned by the `format()` method,
 1. If the call fails or `strval` is not a string:
    1. Set `strval` to be the concatenation of `{`, `mv.source`, and `}`.
-1. Append `strval` to the end of `res`.
+   1. Set `dir` to be `"auto"`.
```
It's re-set here because the failed part is formatted using fallback, which means that its formatted contents are now `{source}` rather than whatever they were supposed to be. This means that since the `source` can be either LTR or RTL, we don't know its direction and need to use `'auto'`.

aphillips replied:
Hm... is `source` that message pattern string? Or externally provided (such as a resource key)?

eemeli replied:
It's the _fallback value_ defined in the MF2 spec: https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#fallback-resolution
